### PR TITLE
Image Importer Improvements

### PR DESCRIPTION
**Pull request by ErisDS**
ref #4608, #4609, #4690
- fix errors with cleaning up files
- improve handling of base directories, and introduce a simple valid format for zips (must contain importable files or folders, and may contain up to one base directory)
- vastly improve test coverage

### CODE REVIEW

**Comment by ErisDS** [comment]
Note: This PR comes before #4692, as it is a follow up to #4690 - will rebase the markdown one on this one.

File: core/server/data/importer/index.js [hunk 1]
```diff
@@ -99,12 +174,38 @@ _.extend(ImportManager.prototype, {
...
+        if (extMatches.length > 0 || dirMatches.length > 0) {
+            return Promise.resolve();
+        }
+
+        // There is a base directory, grab it from any ext match
+        extMatchesAll = glob.sync(
+            this.getExtensionGlob(this.getExtensions(), ALL_DIRS), {cwd: directory}
+        );
+        if (extMatchesAll.length < 1 || extMatchesAll[0].split('/') < 1) {
+            return Promise.resolve(new errors.ValidationError('Invalid zip file: base directory read failed'));
```
Comment (jaswilli) [inline core/server/data/importer/index.js]: I think this should be a `reject()`?  It's hiding a test failure.

File: core/test/unit/importer_spec.js [hunk 2]
```diff
@@ -59,13 +77,11 @@ describe('Importer', function () {
...
-                    fileSpy = sandbox.stub(ImportManager, 'processFile').returns(Promise.resolve()),
-                    cleanSpy = sandbox.stub(ImportManager, 'cleanUp').returns(Promise.resolve());
+                    fileSpy = sandbox.stub(ImportManager, 'processFile').returns(Promise.resolve());
 
                 ImportManager.loadFile(testFile).then(function () {
                     zipSpy.calledOnce.should.be.false;
                     fileSpy.calledOnce.should.be.true;
-                    cleanSpy.calledOnce.should.be.true;
                     done();
                 });
```
Comment (jaswilli) [inline core/test/unit/importer_spec.js]: There are a bunch of these test blocks that are missing `.catch(done)` handlers.

**Comment by jaswilli** [comment]
As I was going through this I noticed a couple methods in `importer/index.js` that are returning promises when they don't need to (`isValidZip` and `getBaseDirectory`).  If they're switched to returning values/throwing errors it can make `processZip` and `loadFile` a little more straightforward and efficient.

I put up the changes here: https://github.com/jaswilli/Ghost/commit/ec6698fa720983fad7b25af5abe29096ac4f448b

Use or ignore :smile:.  I know you're trying to get to a destination that's further down the road than this PR--just playing around with this and thought I'd toss it out there.

File: core/server/data/importer/index.js (hunk 1 above)
Comment (ErisDS) [inline core/server/data/importer/index.js]: yep, whoops

**Comment by ErisDS** [comment]
Looks good to me - at one point I had branching code where the promises made more sense - but this is tonnes better now, just cherry picked your work into the PR :stuck_out_tongue_winking_eye:

**Pull request merged**